### PR TITLE
Bind 'F' key to flood fill for labels layers

### PR DESCRIPTION
**Pull request by GenevieveBuckley**
# Description
This PR makes the experience of annotating labels layers much nicer.

It adds a keybinding where you hold "F" to activate flood fill mode, and automatically deactivates when you release the key. 

This is similar to the keybinding where you hold "Space"r to activate pan-zoom mode, which also automatically de-activates when you release the key.

### Advantages

The benefits of this PR are:
1. A simpler, less annoying workflow when annotating labels, and
2. It conforms more with the behaviour of other image software (Paintera using "F" to fill, so does the Microscopy Image Browser, etc.) This makes it more natural for users to remember and use, especially if they are familiar with other tools.

### An example
Previously, if you were annotating a lot of blobs with different label ids on an image, you 
1. Select paint mode (hotkey shortcut "2")
3. Draw the outline of your object (click and drag with the left mouse button)
4. Select flood fill mode (hotkey shortcut "3")
5. Click inside the outline you have just drawn
6. Increment the label id (hotkey shortcut "+")
... and continue from step one again.

With this PR, assuming you are already starting in paint mode you can:
1. Draw the outline of your object (click and drag with the left mouse button)
2. Hold the "F" key down and click the region you want to fill
3. Increment the label id (hotkey shortcut "+")
... and continue from step one again. I find this is a lot more natural to use in practice.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How has this been tested?

- [x] I've added a new test to `napari/layers/labels/_tests/test_labels_keybindings.py`, where the test structure is modeled off the pan-zoom keybinding tests used for other layers.

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [?] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by GenevieveBuckley** [comment]
Just to check, I don't need to add anything else to the documentation, do I?  I'm assuming this is pulled automatically from the docstring.

I also don't think I need translation strings for this case either, but if I'm wrong will somebody please let me know.

**Comment by psobolewskiPhD** [comment]
Love this!
Quick toggle (hold/release) key binds are super!
While the 12345 key binds are nice, losing the toggles was a bummer (PR: https://github.com/napari/napari/pull/3072) 
Relevant Zulip: https://napari.zulipchat.com/#narrow/stream/212875-general/topic/Labels.20layer.20info.2Fkeybinds/near/259723938

In the end, it's probably impossible to satisfy everyones preferred keybinds, but are these sorts of toggles able to be set in the preferences? I think not:
image
Would love for the original `switch` keybinds to be available—even if unbound by default.

File: napari/layers/labels/_tests/test_labels_key_bindings.py
Comment (Czaki) [inline napari/layers/labels/_tests/test_labels_key_bindings.py]:
```suggestion
    gen = key_bindings.hold_to_pan_zoom(layer)
    next(gen)
    assert layer.mode == 'pan_zoom'
    with contextlib.suppress(StopIteration):
        next(gen)
    assert layer.mode == 'paint'
```

**Review by Czaki — CHANGES_REQUESTED** [review]
I like this idea but I would like to see better tests

File: napari/layers/labels/_tests/test_labels_key_bindings.py
Comment (Czaki) [inline napari/layers/labels/_tests/test_labels_key_bindings.py]:
```suggestion
    gen = key_bindings.hold_to_flood_fill(layer)
    next(gen)
    assert layer.mode == 'fill'
    with contextlib.suppress(StopIteration):
        next(gen)
    assert layer.mode == 'paint'
```

**Comment by Czaki** [comment]
> In the end, it's probably impossible to satisfy everyones preferred keybinds, but are these sorts of toggles able to be set in the preferences? I think not:

Currently, it is not easily possible. One could add functions with `Shift` shortcuts, but then will be no synchronization between toggle and normal shortcuts. To do this there will need to improve shortcut manager to be possible to define the mode of working.

**Comment by Czaki** [comment]
Small observation (I think that is out of the scope of this PR). If You select "paint", hold `F`, hold `Space` release `F` (current mode is "paint"), release `Space`, and end with `fill` mode.

A little unexpected, but may need a big refactor to prevent.

**Comment by jni** [comment]
@GenevieveBuckley what do you think about holding the fill shortcut (3) rather than F? That is, we use the same key for switching and for holding, the difference being the duration and/or whether you click during the hold.

**Comment by GenevieveBuckley** [comment]
> @GenevieveBuckley what do you think about holding the fill shortcut (3) rather than F? That is, we use the same key for switching and for holding, the difference being the duration and/or whether you click during the hold.

No, because then you would lose the ability to set the fill mode, it would disappear anytime the "3" key was released. There's no distinction between a short keypress and a long key hold, as far as I'm aware.

I also don't love it because this is then a totally napari specific UI, and doesn't match what is most common across other software packages.

**Comment by GenevieveBuckley** [comment]
@psobolewskiPhD 

> In the end, it's probably impossible to satisfy everyones preferred keybinds...

Well sure, but I think we can at least aim for being roughly similar to the consensus for certain actions among other visualization software. 

> are these sorts of toggles able to be set in the preferences? I think not:
image

Is this a request for this PR? I can't tell.
I'm not sure how that preferences list is being populated, so I guess we could, but I'm not sure how.

**Comment by GenevieveBuckley** [comment]
@Czaki 
> I like this idea but I would like to see better tests

If you like, but you will need to put in a separate PR to address every other place with tests like these. I copied this format from the shapes keybindings tests, and I think it exists in several other places too.

> Small observation (I think that is out of the scope of this PR). If You select "paint", hold `F`, hold `Space` release `F` (current mode is "paint"), release `Space`, and end with `fill` mode.
> 
> A little unexpected, but may need a big refactor to prevent.

This seems convoluted, and not super likely to happen by accident. It's also easy enough to correct if you do find yourself in that situation, by pressing "2" or clicking the paint mode icon again. For both those reasons, I don't think it's worth investing a lot of time in this particular edge case.

**Comment by jni** [comment]
> No, because then you would lose the ability to set the fill mode, it would disappear anytime the "3" key was released. There's no distinction between a short keypress and a long key hold, as far as I'm aware.

No, this is incorrect. Even if technically it's not possible to directly distinguish between press and hold, you can certainly do it in code — start a timer each time mode changes (which is done on key press), and if the time on key release is less than the "hold" threshold *and* there haven't been any click actions, don't switch back.

> I also don't love it because this is then a totally napari specific UI, and doesn't match what is most common across other software packages.

I think this is a separate issue tbh: *which* key should represent fill (and others). It's open to debate. I'll note that we *did* use F for fill but then @haesleinhuepf opened #2800 where I think he makes a very good point. Separately @AbigailMcGovern had already hacked the viewer to achieve the same thing locally because she was sick of hunt-and-pecking for the right shortcut, compared with just resting her hand on a row of keys.

My proposal is:

- same key for press and for hold, using the hacks described above if necessary.
- that key is controlled by the preference. Then it's easy to accommodate both approaches, and we can have a discussion about what the default should be.
- *potentially*, ability to set more than one key to the same action. Then we could have *both* F and 3 for fill.
- "preference groups", where people could switch between "ergonomic" layer control keybindings and "mneumonic" layer control keybindings, and maybe both.

**Comment by Czaki** [comment]
> If you like, but you will need to put in a separate PR to address every other place with tests like these. I copied this format from the shapes keybindings tests, and I think it exists in several other places too.

"We will not test new code in a proper way because in other places we do not test similar code in a proper way" is not a valid argument. Especially when a better way of testing is suggested.

If the problem is wider then the suggested (or other) code could be generalized to function/fixture for better reuse in the following PR.

> _potentially_, ability to set more than one key to the same action. Then we could have _both_ F and 3 for fill.

there is such ability, but only from the console. It needs to be exposed to the interface only. 

> start a timer each time mode changes

do not start a timer, calculate the difference between press and release.

```python
@Labels.bind_key('Space')
def hold_to_pan_zoom(layer: Labels):
    """Hold to pan and zoom in the viewer."""
    if layer._mode != Mode.PAN_ZOOM:
        # on key press
        prev_mode = layer.mode
        layer.mode = Mode.PAN_ZOOM
        start_time = time.time()

        yield
        
        if time.time() - start_time < get_settings().application.threshold_for_call:
            # on key release
            layer.mode = prev_mode
```

**Comment by jni** [comment]
> "We will not test new code in a proper way because in other places we do not test similar code in a proper way" is not a valid argument.

I think you could be a bit more gentle here @Czaki. Generally, our policy with PRs has been that if they improve the library, then additional improvements *beyond the current state of the library* are more or less optional, depending on the experience and time available of the contributor.

Having said this, @GenevieveBuckley, I do agree with @Czaki that if the code is already written, I don't see a reason not to commit @Czaki's suggestions, unless you see a specific problem with that code. (I don't see one a priori, but I have not pulled it down and tested it.) Or maybe you missed @Czaki's [suggested changes](https://github.com/napari/napari/pull/4696#discussion_r897346184) altogether?

**Comment by GenevieveBuckley** [comment]
@jni I think you have misunderstood. Of course I will commit @Czaki's suggested changes. I didn't do that before because it was very late at night (perhaps you didn't realize that?). *EDIT: Oh, I see @Czaki has also possibly misunderstood. Look everybody, it is just not reasonable to expect people to be doing work at 10pm, please hold off on any complaints about perceived failures to address suggestions for at least 24-48 hours.*

And @Czaki will **also** need to make a separate PR to address the same concern in all the other places this test pattern is already repeated elsewhere in napari. I assume it's not immediately clear to everyone that those places also exist, which is why I commented about the need for a second PR above.

**Comment by jni** [comment]
@GenevieveBuckley yeah async communication is hard. No one was expecting you to work at any particular hour — but your response sounded like you didn't want to fix the tests *even before the PR merged.* All good now. 😊 Well, except everything is failing! 😬 (It was failing before @Czaki's commits too.)

**Comment by Czaki** [comment]
Thear is two reasons for failure. Basic one is that this PR add new layer action but not increase number of expected actions ` >assertlen(layer_methods) == expectation` 

Also there is no import of contextlib in test file

**Comment by GenevieveBuckley** [comment]
CI errors now appear to be failures of xvfb-run

```
 ___________________ summary ___________________
Error: ERROR:   py39-linux-pyside2: commands failed
/usr/bin/bash /home/runner/work/_actions/GabrielBB/xvfb-action/v1/cleanup.sh
Killing the following xvfb processes: 3067
Error: The process '/usr/bin/xvfb-run' failed with exit code 1
```

**Comment by Czaki** [comment]
I see these errors already. There is some problem with packages updated since last passing tests.

**Comment by jni** [comment]
Looks like a change in the Ubuntu runner, based on seeing this VisPy PR in my feed:

https://github.com/vispy/vispy/pull/2356

**Comment by brisvag** [comment]
I opened tlambert03/setup-qt-libs#57 to address that. Hopefully that's correct :P

**Review by Czaki — APPROVED** [review]

**Comment by GenevieveBuckley** [comment]
> I opened [tlambert03/setup-qt-libs#57](https://github.com/tlambert03/setup-qt-libs/pull/57) to address that. Hopefully that's correct :P

Tests passing! Thanks for the fix @brisvag

**Comment by GenevieveBuckley** [comment]
Spoke too soon, I guess. The problem with the CI is still ongoing. @brisvag do you think this is the same issue or a different one?

**Comment by brisvag** [comment]
Nope, this looks like a real failure. You need to increase the `EXPECTED_NUMBER_OF_LAYER_METHODS` for `Labels`, it seems :) (somehow?)

**Comment by GenevieveBuckley** [comment]
> Nope, this looks like a real failure. You need to increase the `EXPECTED_NUMBER_OF_LAYER_METHODS` for `Labels`, it seems :) (somehow?)

I thought I did! Don't tell me that's somehow been overwritten by the last merge with main...

**Comment by GenevieveBuckley** [comment]
So [this is the commit](https://github.com/napari/napari/pull/4696/commits/35f5044178a025beeeb3719005b6f641b94b6133) where I increased the `EXPECTED_NUMBER_OF_LAYER_METHODS`.

And these are the error messages I'm seeing from the continuous integration. This one is from the latest CI run, see here: https://github.com/napari/napari/runs/7459687387?check_suite_focus=true
```
Run GabrielBB/xvfb-action@v1
/usr/bin/sudo apt-get install -y xvfb
Reading package lists...
Building dependency tree...
Reading state information...
xvfb is already the newest version (2:1.20.13-1ubuntu1~20.04.3).
0 upgraded, 0 newly installed, 0 to remove and 18 not upgraded.
Command: python -m tox
/usr/bin/xvfb-run --auto-servernum python -m tox
.package create: /tmp/.tox/.package
.package installdeps: setuptools >= 42, wheel, setuptools_scm[toml]>=3.4
/tmp/.tox/.package/lib/python3.8/site-packages/setuptools/config/setupcfg.py:463: SetuptoolsDeprecationWarning: The license_file parameter is deprecated, use license_files instead.
  warnings.warn(msg, warning_class)
/tmp/.tox/.package/lib/python3.8/site-packages/setuptools_scm/git.py:135: UserWarning: "/home/runner/work/napari/napari" is shallow and may cause errors
  warnings.warn(f'"{wd.path}" is shallow and may cause errors')
/tmp/.tox/.package/lib/python3.8/site-packages/setuptools/config/setupcfg.py:463: SetuptoolsDeprecationWarning: The license_file parameter is deprecated, use license_files instead.
  warnings.warn(msg, warning_class)
warning: no previously-included files found matching 'napari/resources/_qt_resources*.py'
warning: no previously-included files matching '*.pyc' found under directory 'napari'
/tmp/.tox/.package/lib/python3.8/site-packages/setuptools_scm/git.py:135: UserWarning: "/home/runner/work/napari/napari" is shallow and may cause errors
  warnings.warn(f'"{wd.path}" is shallow and may cause errors')
/tmp/.tox/.package/lib/python3.8/site-packages/setuptools/config/setupcfg.py:463: SetuptoolsDeprecationWarning: The license_file parameter is deprecated, use license_files instead.
  warnings.warn(msg, warning_class)
warning: no previously-included files found matching 'napari/resources/_qt_resources*.py'
warning: no previously-included files matching '*.pyc' found under directory 'napari'
py38-linux-pyqt5 create: /tmp/.tox/py38-linux-pyqt5
tox: py38-linux-pyqt5
___________________________________ summary ____________________________________
Error: ERROR:   py38-linux-pyqt5: commands failed
/usr/bin/bash /home/runner/work/_actions/GabrielBB/xvfb-action/v1/cleanup.sh
No xvfb processes to kill
Error: The process '/usr/bin/xvfb-run' failed with exit code 1
```

**Comment by brisvag** [comment]
Uhm, but if you open the `tox` dropdown the error is this right?

```py
  _______________ test_expected_number_of_layer_methods[Points-8] ________________
  
  cls = 'Points', expectation = 8
  
  @pytest.mark.parametrize(
  'cls, expectation', EXPECTED_NUMBER_OF_LAYER_METHODS.items()
      )
  deftest_expected_number_of_layer_methods(cls, expectation):
  """
      Make sure we do find all the methods attached to a layer via keybindings
      """
          layer_methods = _get_all_keybinding_methods(getattr(layers, cls))
  >       assertlen(layer_methods) == expectation
  E       AssertionError: assert 9 == 8
  E        +  where 9 = len({'activate_points_add_mode', 'activate_points_pan_zoom_mode', 'activate_points_select_mode', 'copy', 'delete_selected_points', 'hold_to_pan_zoom', ...})
  
  napari/_tests/test_viewer.py:117: AssertionError
```

Apparently something is off for the `Points` layer O.o

File: napari/_tests/test_viewer.py
Comment (Czaki) [inline napari/_tests/test_viewer.py]: here is the decrease of points number of expected methods. I do not know why.

Comment (psobolewskiPhD) [reply napari/_tests/test_viewer.py]: A recent PR (https://github.com/napari/napari/pull/4809) added a points-layer short cut and increased the number of methods to 9, from the original 8. It was merged a few days ago, so perhaps this issue is due to a merge conflict?

**Comment by jni** [comment]
@GenevieveBuckley sorry we didn't explicitly ping here earlier, but @Czaki implemented the "short press/long hold" pattern in #4814. I would say that that PR in conjunction with keybinding preferences are enough to meet your needs? ie maybe it's not worth spending more effort fixing the CI here.

File: napari/_tests/test_viewer.py
Comment (GenevieveBuckley) [inline napari/_tests/test_viewer.py]:
```suggestion
    'Points': 9,
```

File: napari/_tests/test_viewer.py
Comment (GenevieveBuckley) [reply napari/_tests/test_viewer.py]: Ooh yeah, I haven't touched the points layer at all. That does look like a problem caused by an improper merge - thanks!

**Comment by Czaki** [comment]
As https://github.com/napari/napari/pull/4814 is merged there, a need to be made the decision about this PR.

**Comment by GenevieveBuckley** [comment]
There hasn't been much discussion here (or anywhere else I know of) about identifying and adhering to a set of user interface controls common across multiple software types, which is disappointing.

I did bring this up in the last meeting, but without much progress. For those who weren't at the meeting, you can search for "user interaction design" in the meeting notes for a short summary.

**Comment by GenevieveBuckley** [comment]
And given that it literally says this in the napari docs:
> You can quickly select the fill bucket by pressing the `F` key when the labels layer is selected.

https://github.com/napari/napari/blob/main/docs/howtos/layers/labels.md?plain=1#L261-L262

I think it's probably better to have this use the same mechanism as in the other PR (short press to switch modes, long hold to toggle mode on/off), both for simplicity's sake and ease of use.

**Comment by Czaki** [comment]
Ahh. Again lack of synchronization between documentation and code. we should put comments like: "If you change these please update this file with documentation".

**Comment by Czaki** [comment]
What with this PR?

**Comment by psobolewskiPhD** [comment]
Seems like now with all the keybind changes in 0.4.17RC (ie. 2ndary keybinds possible), hard-coding `F` isn't needed? 
We can just pre-bind the 2ndary keybind here:
https://github.com/napari/napari/blob/8bb7e78f8fa9ec05fa82302c9e27254a85c1cc11/napari/utils/shortcuts.py#L17
Should be the same effect?

**Comment by Czaki** [comment]
Yes. There is a question if this PR should be closed or somehow changed.

**Comment by brisvag** [comment]
> Yes. There is a question if this PR should be closed or somehow changed.

Either is totally fine :)

**Comment by GenevieveBuckley** [comment]
Superseded by https://github.com/napari/napari/pull/5155

**Pull request closed without merging**
